### ISSUE TITLE
[Help Wanted] FP16 support

### ISSUE THREAD

**Issue by vince62s** [OpenNMT/OpenNMT-py#1099]
There is a howto here https://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html#pytorch

Should not be very difficult to implement, maybe tricky to tune with a good scale_factor.

Contribution welcome.

we can expect a +25% training speed gain on a RTX 2080ti.

**Comment by guillaumekln** [comment]
Interested in looking into this. Did someone already experiment something?

**Comment by vince62s** [comment]
You're very welcome!
we did not much. we changed a few lines to make it "run" and see the speed on RTX but not functional since more needs to be done. (fp32 softmax, fp6 inputs, ...)

diff --git a/onmt/modules/multi_headed_attn.py b/onmt/modules/multi_headed_attn.py
index 97b7961..83ca3e7 100644
--- a/onmt/modules/multi_headed_attn.py
+++ b/onmt/modules/multi_headed_attn.py
@@ -178,7 +178,7 @@ class MultiHeadedAttention(nn.Module):
 
         if mask is not None:
             mask = mask.unsqueeze(1).expand_as(scores)
-            scores = scores.masked_fill(mask, -1e18)
+            scores = scores.masked_fill(mask, -float('inf'))
 
         # 3) Apply attention dropout and compute context vectors.
         attn = self.softmax(scores)
diff --git a/onmt/trainer.py b/onmt/trainer.py
index 09bd8c2..f5d1cbc 100644
--- a/onmt/trainer.py
+++ b/onmt/trainer.py
@@ -109,6 +109,8 @@ class Trainer(object):
                 """To enable accumulated gradients,
                    you must disable target sequence truncating."""
 
+       self.model.half()
+
         # Set model in training mode.
         self.model.train()
 
diff --git a/onmt/utils/loss.py b/onmt/utils/loss.py
index 623692a..b11f274 100644
--- a/onmt/utils/loss.py
+++ b/onmt/utils/loss.py
@@ -200,7 +200,7 @@ class LabelSmoothingLoss(nn.Module):
         model_prob.scatter_(1, target.unsqueeze(1), self.confidence)
         model_prob.masked_fill_((target == self.padding_idx).unsqueeze(1), 0)
 
-        return F.kl_div(output, model_prob, reduction='sum')
+        return F.kl_div(output.float(), model_prob, reduction='sum')
